core/src/main/java/org/apache/carbondata/core/datastore/chunk/store/impl/unsafe/UnsafeVariableLengthDimensionDataChunkStore.java

```diff
@@ -78,70 +88,96 @@ public UnsafeVariableLengthDimensionDataChunkStore(long totalSize, boolean isInv
-      // incrementing the pointers as first value is already filled and as we are storing as int
-      // we need to increment the 4 bytes to set the position of the next value to set
-      pointerOffsets += CarbonCommonConstants.INT_SIZE_IN_BYTE;
+      dataOffsets[i] = startOffset + CarbonCommonConstants.SHORT_SIZE_IN_BYTE;
     }
-
+    CarbonUnsafe.getUnsafe().copyMemory(dataOffsets, CarbonUnsafe.INT_ARRAY_OFFSET,
+        dataPageMemoryBlock.getBaseObject(),
+        dataPageMemoryBlock.getBaseOffset() + this.dataPointersOffsets,
+        dataOffsets.length * CarbonCommonConstants.INT_SIZE_IN_BYTE);
   }
 
   /**
    * Below method will be used to get the row based on row id passed
-   *
+   * Getting the row from unsafe works in below logic
+   * 1. if inverted index is present then get the row id based on reverse inverted index
+   * 2. get the current row id data offset
+   * 3. if it's not a last row- get the next row offset
+   * Subtract the current row offset + 2 bytes(to skip the data length) with next row offset
+   * 4. if it's last row
+   * subtract the current row offset + 2 bytes(to skip the data length) with complete data length
    * @param rowId
    * @return row
    */
   @Override public byte[] getRow(int rowId) {
+    // get the actual row id
+    rowId = getRowId(rowId);
+    // get offset of data in unsafe
+    int currentDataOffset = getOffSet(rowId);
+    // get the data length
+    short length = getLength(rowId, currentDataOffset);
+    // create data array
+    byte[] data = new byte[length];
+    // fill the row data
+    fillRowInternal(length, data, currentDataOffset);
+    return data;
+  }
+
+  /**
+   * Returns the actual row id for data
+   * if inverted index is present then get the row id based on reverse inverted index
+   * otherwise return the same row id
+   * @param rowId
+   * row id
+   * @return actual row id
+   */
+  private int getRowId(int rowId) {
     // if column was explicitly sorted we need to get the rowid based inverted index reverse
     if (isExplicitSorted) {
       rowId = CarbonUnsafe.getUnsafe().getInt(dataPageMemoryBlock.getBaseObject(),
           dataPageMemoryBlock.getBaseOffset() + this.invertedIndexReverseOffset + ((long)rowId
               * CarbonCommonConstants.INT_SIZE_IN_BYTE));
     }
-    // now to get the row from memory block we need to do following thing
-    // 1. first get the current offset
-    // 2. if it's not a last row- get the next row offset
-    // Subtract the current row offset + 2 bytes(to skip the data length) with next row offset
-    // else subtract the current row offset + 2 bytes(to skip the data length)
-    // with complete data length
-    int currentDataOffset = CarbonUnsafe.getUnsafe().getInt(dataPageMemoryBlock.getBaseObject(),
-        dataPageMemoryBlock.getBaseOffset() + this.dataPointersOffsets + (rowId
+    return rowId;
+  }
+
+  /**
+   * get data offset based on current row id
+   * @param rowId
+   * row id
```

Comment:
move it to previous line, please modify all places

Reply:
ok